File: mcp_server_development/python/src/session14/answers/q4_cli_diff.py
from __future__ import annotations

import argparse
import os

DEFAULT_MAX_RESULTS = 5
MAX_ALLOWED_RESULTS = 20


def bounded_int(raw: str) -> int:
    """argparse の type に渡す変換関数。

    ここで ArgumentTypeError を投げると、argparse が使い方を表示して
    終了コード 2 で落ちてくれます（TypeScript 版の process.exit(2) と同じ意味）。
    """
    try:
        value = int(raw)
    except ValueError:
        raise argparse.ArgumentTypeError(f"整数を指定してください（受け取った値: {raw}）") from None
    if not 1 <= value <= MAX_ALLOWED_RESULTS:
        raise argparse.ArgumentTypeError(
            f"1 以上 {MAX_ALLOWED_RESULTS} 以下で指定してください（受け取った値: {value}）"
        )
    return value


# build_parser() に追加する。既定値を環境変数から取るのが要点
#   環境変数の検証も bounded_int に通したいので、default ではなく後段で解決する
def add_max_results(parser: argparse.ArgumentParser) -> None:
    parser.add_argument(
        "--max-results",
        dest="max_results",
        type=bounded_int,
        default=None,
        metavar="N",
        help=f"検索結果の既定件数（1〜{MAX_ALLOWED_RESULTS}、既定 {DEFAULT_MAX_RESULTS}）",
    )


def resolve_max_results(cli_value: int | None, env: dict[str, str] | None = None) -> int:
    environ = os.environ if env is None else env
    if cli_value is not None:
        return cli_value
    raw = (environ.get("DOCSEARCH_MAX_RESULTS") or "").strip()
    if not raw:
        return DEFAULT_MAX_RESULTS
    # 環境変数も引数と同じ検証を通す。経路によって厳しさが変わってはいけない
    return bounded_int(raw)

File: mcp_server_development/node/src/session14/answers/q4-cli-diff.ts
/** 検索件数の既定値と上限。ドメイン層の MAX_LIMIT と揃えておく */
const DEFAULT_MAX_RESULTS = 5;
const MAX_ALLOWED_RESULTS = 20;

type MaxResults =
  | { readonly ok: true; readonly value: number }
  | { readonly ok: false; readonly message: string };

/**
 * ① 引数 --max-results → ② 環境変数 DOCSEARCH_MAX_RESULTS → ③ 既定値 5
 *
 * 環境変数は必ず文字列で届きます。数値に見えない値・範囲外の値は
 * 「無言で既定値に戻す」のではなく、起動時のエラーにします。
 * 無言で戻すと、利用者は「設定したのに効かない」を延々と疑うことになります。
 */
function resolveMaxResults(cliValue: string | undefined, env: NodeJS.ProcessEnv): MaxResults {
  const raw = (cliValue ?? env["DOCSEARCH_MAX_RESULTS"] ?? "").trim();
  if (raw.length === 0) {
    return { ok: true, value: DEFAULT_MAX_RESULTS };
  }
  if (!/^[0-9]+$/.test(raw)) {
    return { ok: false, message: `--max-results には整数を指定してください（受け取った値: ${raw}）` };
  }
  const value = Number.parseInt(raw, 10);
  if (value < 1 || value > MAX_ALLOWED_RESULTS) {
    return {
      ok: false,
      message: `--max-results は 1 以上 ${MAX_ALLOWED_RESULTS} 以下で指定してください（受け取った値: ${value}）`,
    };
  }
  return { ok: true, value };
}

// parseArgs() の分岐に追加する（--max-results 3 と --max-results=3 の両方を受ける）
// } else if (token === "--max-results") {
//   const value = argv[i + 1];
//   if (value === undefined || value.startsWith("-")) {
//     invalid ??= "--max-results には値が必要です";
//   } else { maxResults = value; i += 1; }
// } else if (token.startsWith("--max-results=")) {
//   maxResults = token.slice("--max-results=".length);

// 起動処理の末尾に追加する（cli.ts 側に書く。このファイル単体では options / pkg が無いので
// コメントにしてあります。そのまま cli.ts に移してください）
//
// const maxResults = resolveMaxResults(options.maxResults, process.env);
// if (!maxResults.ok) {
//   process.stderr.write(`${pkg.name}: ${maxResults.message}\n`);
//   process.exit(2);
// }
// process.stderr.write(
//   `[${pkg.name}] v${pkg.version} / docsRoot=${docsRoot} / maxResults=${maxResults.value}\n`,
// );
